### ISSUE TITLE
Training Collapse in Stage 2 (Prior Training): Metrics Drop Significantly and Generated Images Turn Black

### ISSUE THREAD

**Issue by yuanfeilitianzhe** [xmed-lab/ZEBRA#6]
I am writing to report a training collapse issue I encountered during Stage 2 (Diffusion Prior training). While Stage 1 (Pretraining) completes successfully, transitioning to Stage 2 (pretrain=False) leads to a sudden and severe drop in evaluation metrics after a few epochs (around Epoch 3), despite the training loss continuing to decrease.

Experimental Observations：
1、Performance Drop: In Epoch 2, the model performed well (e.g., test_fwd_percent_correct: 0.854). However, starting from Epoch 3, the metrics crashed (e.g., test_fwd_percent_correct: 0.069), and by Epoch 4, test_blurry_pixcorr became nan.
2、Visual Output: I attempted to generate images using the checkpoint from the collapsed epoch, and the results were completely black images. This explains why pixcorr resulted in nan.

Have you observed similar collapse issues during Stage 2 training?

I would appreciate any insights or suggestions you might have. Thank you for your time and help!

The following is my log.
Training Epoch 2 Loss Summary ... 
	loss: 21.226252 
	loss_prior: 0.507851 
	loss_clip_vision: 2.646981 
	loss_clip_vision_s: 3.726115 
	loss_image_cls_all: 0.040854 
	loss_image_cls: 0.053173 
	loss_image_dis: 0.041898 
	loss_clip_txt: 2.134746 
	loss_blurry: 0.583842 
Evaluating Epoch 2 ... 
	test_fwd_percent_correct: 0.8540000116825104 
	test_bwd_percent_correct: 0.7970000159740448 
	text_fwd_percent_correct: 0.8520000350475311 
	test_blurry_pixcorr     : 0.47802490234375 
New best test metric: 2.503000062704086 
---saved <CHECKPOINT_DIR>/brain_model_prior ckpt!---
---saved <CHECKPOINT_DIR>/brain_model_prior_last ckpt!---
[2026-03-26 11:15:32,155] [INFO] [unfused_optimizer.py:289:_update_scale] No Grad overflow for 1000 iterations
[2026-03-26 11:15:32,156] [INFO] [unfused_optimizer.py:290:_update_scale] Increasing dynamic loss scale from 16384.0 to 32768.0

Training Epoch 3 Loss Summary ... 
	loss: 19.840729 
	loss_prior: 0.465668 
	loss_clip_vision: 2.533575 
	loss_clip_vision_s: 3.700416 
	loss_image_cls_all: 0.038570 
	loss_image_cls: 0.053001 
	loss_image_dis: 0.040321 
	loss_clip_txt: 2.027225 
	loss_blurry: 0.568145 
Evaluating Epoch 3 ... 
	test_fwd_percent_correct: 0.0690000007301569 
	test_bwd_percent_correct: 0.5020000112056732 
	text_fwd_percent_correct: 0.2739999997615814 
	test_blurry_pixcorr     : 0.5135791015625 
Current metric: 0.8450000116974115, best metric loss is 2.503000062704086 in Epoch 2 
---saved <CHECKPOINT_DIR>/brain_model_prior_last ckpt!---
[2026-03-26 11:42:34,297] [INFO] [unfused_optimizer.py:289:_update_scale] No Grad overflow for 1000 iterations
[2026-03-26 11:42:34,298] [INFO] [unfused_optimizer.py:290:_update_scale] Increasing dynamic loss scale from 32768.0 to 65536.0

Training Epoch 4 Loss Summary ... 
	loss: 18.957664 
	loss_prior: 0.439534 
	loss_clip_vision: 2.464759 
	loss_clip_vision_s: 3.684419 
	loss_image_cls_all: 0.036654 
	loss_image_cls: 0.052507 
	loss_image_dis: 0.038225 
	loss_clip_txt: 1.867449 
	loss_blurry: 0.557864 
Evaluating Epoch 4 ... 
	test_fwd_percent_correct: 0.0590000007301569 
	test_bwd_percent_correct: 0.208000003695488 
	text_fwd_percent_correct: 0.265 
	test_blurry_pixcorr     : nan 
Current metric: 0.5320000044256449, best metric loss is 2.503000062704086 in Epoch 2

**Comment by McGregorWwww** [comment]
Sorry, I never encountered this. It seems like it's a DeepSpeed-related issue; the gradients are likely becoming extremely small (not overflowing but underflowing). The ever-increasing loss scale eventually pushes intermediate activations or the prior's predictions into a degenerate regime — the model learns to output near-zero (black images), which can still reduce certain loss terms. Once blurry_pixcorr hits NaN, it confirms the outputs have zero variance.

Here are some workarounds that may help:
1. Lower the learning rate for Stage 2.
2. Add gradient clipping. If you aren't already clipping, set gradient_clipping to something like 1.0 in your DeepSpeed config.
3. Cap or reduce the loss scale. Set initial_scale_power lower (e.g., 12 instead of the default) and set a loss_scale_window or max_scale in your DeepSpeed FP16 config to prevent unbounded scaling. Alternatively, switch to BF16 if your hardware supports it — it avoids the dynamic scaling problem entirely.
